rest-api/dao/questionnaire_response_dao.py

```diff
@@ -83,15 +83,51 @@ def _validate_model(self, session, obj):  # pylint: disable=unused-argument
       logging.error(
           'QuestionnaireResponse model has no answers. This is harmless but probably an error.')
 
+  def _validate_link_ids_from_resource_json_group(self, resource, link_ids):
+    """
+    Look for question sections and validate the linkid in each answer. If there is a response
+    answer link id that does not exist in the questionnaire, then raise an exception.
+    This is a recursive function because answer groups can be nested.
+    :param resource: A group section of the response json.
+    :param link_ids: List of link ids to validate against.
+    """
+    # look for child group sections and call ourselves
+    if 'group' in resource:
+      self._validate_link_ids_from_resource_json_group(resource['group'], link_ids)
+
+    if type(resource) is list:
+      for item in resource:
+        self._validate_link_ids_from_resource_json_group(item, link_ids)
+
+    # look for question section:
+    if 'question' in resource:
+
+      # if question is a list, then loop through the list
+      if type(resource['question']) is list:
+        self._validate_link_ids_from_resource_json_group(resource['question'], link_ids)
+
+      # if question is not a list, check linkId directly
```

Comment:
Not following here, this claims question is not a list (but we didn't return in the above if statement L107, so this may not be true).

Then we iterate through the question anyways; are we assuming it's a dict instead? Should probably make this check explicit.

Reply:
Sorry - I still don't follow. The comment says "question is not a list", you then proceed to iterate through the question.

The comment also implies exclusivity with line 107 which isn't true because you don't return at L107.